### PR TITLE
Feat: Add Restart Counter Metric for System Components

### PR DESCRIPTION
**Pull request by UlrichSchlueter**
### Which issue this PR addresses:

Feat: Add Restart Counter Metric for System Components

https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9534801

### What this PR does / why we need it:

This PR adds a logic to report pods/containers that saw more than 10 restarts to Geneva. A Geneva alert will be added to create IcMs when the restart rate exceeds a certain limit.

The reporting threashold of 10 is currently hardcoded.

### Test plan for issue:

Manual Unit Test, testing the the code needs a 'broken' cluster that shows the restart behaviour.

### Is there any documentation that needs to be updated for this PR?

not that I'm aware of

### CODE REVIEW

File: pkg/monitor/cluster/podconditions.go
Comment (m1kola) [inline pkg/monitor/cluster/podconditions.go:23]:
```suggestion
var restartCounterThreshold int32 = 10
```

File: pkg/monitor/cluster/podconditions.go
Comment (m1kola) [inline pkg/monitor/cluster/podconditions.go:136]: Maybe we should always send metrics and deal with the alerting logic on the monitor side? That way we can see the full picture somewhere on the dashbaord: not only starting after N restarts. And we will be able to adjust threshold without an RP release. I see that in the [work iterm](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9534801) it was suggested to have this condition in code, but it is not super clear why. Maybe some Geneva limits? @gvanderpotte (and any one else) any thougts?

Also wondering - what is the plan for alerting logic? Create an alert once we have more than N restarts? If so - do you know whether restart count resets at some point? Or is it just being incremented? If I recall correctly, it resets after some time of successfull execution, but I'm not 100% sure. After a quick search I managed to find [this](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy):

> Once a container has executed for 10 minutes without any problems, the kubelet resets the restart backoff timer for that container.

But it is about backoff timer, not about the counter.

File: pkg/monitor/cluster/podconditions.go
Comment (m1kola) [inline pkg/monitor/cluster/podconditions.go:142]: For consistency with other fields.

```suggestion
				"containerName": cs.Name,
```

Same with below in the logger.

File: pkg/util/namespace/namespace.go
Comment (m1kola) [inline pkg/util/namespace/namespace.go:16]: Now we can probably do something like this:

```suggestion
	return ns == "" || ns == "default" || IsOpenShiftSystemNamespace(ns)
```

File: pkg/monitor/cluster/podconditions.go
Comment (petrkotas) [reply pkg/monitor/cluster/podconditions.go:136]: The reason for the threshold is that there ussually is a steady number of restarts after cluster deploy/upgrade. The reason for this is simply the design of the cluster and the eventual consistency. We have often observed around 4 restarts for e.g. for network before it went OK. 

Small amount of restarts ussually does not mean a problem, that was the point. 

As for me, I am OK with the threshold as long as it is kept low (like 5) to simply filter the bootstrapping restarts.

File: pkg/util/namespace/namespace.go
Comment (petrkotas) [inline pkg/util/namespace/namespace.go:23]: This would not be enough. We observe customers installing marketplace stuff e.g. `openshift-gitops`. 

At this moment we do not have better solution, so I am just point that for a long term goal we need to either label namespaces we care about or create denylist for namespaces we do not want.

For the sake of this PR I am happy with this solution.

File: pkg/monitor/cluster/podconditions.go
Comment (m1kola) [reply pkg/monitor/cluster/podconditions.go:136]: @petrkotas yes, it makes sense to have a threshold. The question is - do we put the threshold filtering in the RP (before sending the metrics to Geneva) or in the monitor (when we have metrics in geneva already).

Having threshold sounds like more flexible option: we can adjust threshold without having to release an RP. But maybe there are some implications that I'm missing. What do you think?

File: pkg/monitor/cluster/podconditions.go
Comment (Makdaam) [inline pkg/monitor/cluster/podconditions.go:139]: I'm a bit worried about cardinality of this dimension, but I don't know if there's a better way to distinguish different pod groups in the same namespace. Since we're already using that for other pod.* metrics it's fine here.

Just a note for further discussion.

File: pkg/monitor/cluster/podconditions.go
Comment (Makdaam) [reply pkg/monitor/cluster/podconditions.go:136]: IMO it depends on the Threshold value. 10 seems ok to me.

If we set it too high - there will be a mismatch of expectations for someone looking at the dashboards when they see no data for pods that don't restart, when they're looking for one restart.

If we set it too low - we're going to get a ton of datapoints that aren't useful (I assume 2-3 restarts on a pod are within normal execution flow).

File: pkg/util/namespace/namespace.go
Comment (UlrichSchlueter) [reply pkg/util/namespace/namespace.go:16]: I was planning to do that, but then didn't because I believe the list will probably diverge further.

File: pkg/monitor/cluster/podconditions.go
Comment (UlrichSchlueter) [reply pkg/monitor/cluster/podconditions.go:136]: There was also the point of using a threshold is/was to save on Geneva data costs. 

I don't think the restart counter itself gets reset during the lifetime of a pod (or even beyond https://github.com/kubernetes/kubernetes/issues/54476) . There is a mechanism that pauses restarting after a number of failed attempts in a short amount of time  for some time. 

Having hard-coded values in the metric generation code  does not feel right , but that seems to be the design followed for other metrics, too.

File: pkg/util/namespace/namespace.go
Comment (UlrichSchlueter) [reply pkg/util/namespace/namespace.go:23]: We can add additional filtering on the Geneva side if required, and update this list when see if we get too much noise for customer installed components.

File: pkg/monitor/cluster/podconditions.go
Comment (UlrichSchlueter) [inline pkg/monitor/cluster/podconditions.go:150]:
```suggestion
					"containerName": cs.Name,
```

File: pkg/monitor/cluster/podconditions.go
Comment (UlrichSchlueter) [inline pkg/monitor/cluster/podconditions.go:134]:
```suggestion
			if cs.RestartCount < restartCounterThreshold{
```

File: pkg/monitor/cluster/podconditions.go
Comment (m1kola) [reply pkg/monitor/cluster/podconditions.go]: > Having hard-coded values in the metric generation code does not feel right , but that seems to be the design followed for other metrics, too.

On the metric generation side we normally filter out things like healthy conditions: we only send unhealthy/unexpected conditions. But I don't remember having things like thresholds hardcoded anywhere.

I do not have strong feelings about it: I'm ok with having it here. Having threshold in the monitor side means we send more data, but we are more flexible with the thresholds value (no need to redelpoy RP if we want to tweak it).

However all this probably doesn't really matter on the bigger picture becuase of this:


> I don't think the restart counter itself gets reset during the lifetime of a pod (or even beyond kubernetes/kubernetes#54476) . There is a mechanism that pauses restarting after a number of failed attempts in a short amount of time for some time.

In that case we might have a issue: at what point do we start sending metrics (and more importantly start alerting)? If the restart counter is not being reset - this means that we probably already have some older clusters which are currently healthy but have more than 10 restarts on the counter.

File: pkg/util/namespace/namespace.go
Comment (m1kola) [reply pkg/util/namespace/namespace.go:23]: @petrkotas @UlrichSchlueter I believe @drewandersonnz is going to create a work item for us to loo into filtering out things from customer namespaces which start with `openshift-`, but belong to customers. And @hawkowl also looked into it IIRC.

But yes: this is out of scope of this work item, IMO.

File: pkg/monitor/cluster/podconditions.go
Comment (m1kola) [reply pkg/monitor/cluster/podconditions.go:139]: We can probably drop `nodeName` and/or `containerName` here. Namespace + podname should be enough to identify a problematic container/node.

File: pkg/monitor/cluster/podconditions.go
Comment (UlrichSchlueter) [reply pkg/monitor/cluster/podconditions.go]: The way I understand the ticket, we'd alert on change rate, not total number of restarts.

Afaic we have two phases:
- Data collection: Here the threshold of 10 (or similar) helps reducing the total amount of data we send into Geneva, because we filter out all of the healthy pods on the  RP side.
- Data evaluation: Here we'd look at the rate with which the restarts occur and alert if that rate exceeds another threshold (that we define in Geneva/the monitor). The absolute restart counts would not matter in this case.

I have no good data to suggest a good value for the rate at this point in time. During my tests I did not manage to get above something like 5-10 restarts per hour because of the mentioned protection mechanism, which just makes the cluster stop restarting.

File: pkg/monitor/cluster/podconditions.go
Comment (UlrichSchlueter) [reply pkg/monitor/cluster/podconditions.go:139]: Happy to drop these fields.

File: pkg/monitor/cluster/podconditions.go
Comment (m1kola) [reply pkg/monitor/cluster/podconditions.go]: > Data collection: Here the threshold of 10 (or similar) helps reducing the total amount of data we send into Geneva, because we filter out all of the healthy pods on the RP side.

It becomes less useful given that restart counters do not get reset. But I'm ok with that.

> Data evaluation: Here we'd look at the rate with which the restarts occur and alert if that rate exceeds another threshold (that we define in Geneva/the monitor). The absolute restart counts would not matter in this case.

This makes. Did you have any success defining alerting logic like this in Geneva?

File: pkg/monitor/cluster/podconditions.go
Comment (m1kola) [inline pkg/monitor/cluster/podconditions.go:146]: We are missing namespace here, I think.

**Review by m1kola — COMMENTED** [review]
The code looks good. But I just realised that it is missing unit tests.

---

@UlrichSchlueter as discussed, I'll put the PR on hold until we verify that monitor can be configured to achieve alerting logic we want.

File: pkg/monitor/cluster/podconditions_test.go
Comment (m1kola) [inline pkg/monitor/cluster/podconditions_test.go:145]: It is usually `Test` + `NameOfTheFunctionYouTest`. For example:

* `_emitPodConditions` -> `TestEmitPodConditions`
* `_emitPodContainerStatuses` -> `TestEmitPodContainerStatuses`

```suggestion
func TestEmitPodContainerRestartCounter(t *testing.T) {
```

**Review by m1kola — DISMISSED** [review]
Looks good to me. I think this is ready to merge once we get a confirmation on monitor configuration [as discussed previously](https://github.com/Azure/ARO-RP/pull/1850#pullrequestreview-821458407).

**Review by m1kola — APPROVED** [review]

**Review by hawkowl — APPROVED** [review]

**Pull request merged**
